interfaces/machine_parameters/ring_section.py

```diff
@@ -170,31 +171,37 @@ def __init__(self, length, alpha_0,
         if not isinstance(synchronous_data,
                           ring_programs._synchronous_data_program):
             synchronous_data \
-                = ring_programs._synchronous_data_program._conversions[
-                    func_type](synchronous_data)
+                = getattr(ring_programs, func_type + '_program')(
+                    synchronous_data)
 
         self.synchronous_data = synchronous_data
 
         # Setting orbit length
         if orbit_bump is None:
+
             self.length = ring_programs.orbit_length_program(
                 self.length_design)
             self.orbit_bump = None
+
         else:
+
             if not isinstance(orbit_bump, ring_programs.orbit_length_program):
                 orbit_bump = ring_programs.orbit_length_program(orbit_bump)
 
             self._check_and_set_alpha_and_orbit(orbit_bump)
 
-            # orbit_bump was checked, the length attribute is created
-            # as a copy to keep all datatypes info and length_design
-            # is added onto it
-            self.length = orbit_bump.copy()
-
+            # orbit_bump was checked
+            # The length attribute is created with no further checks
             if orbit_bump.timebase == 'by_time':
-                self.length[:, 1, :] += self.length_design
+
+                self.length = ring_programs.orbit_length_program(
+                    [self.length_design] * len(self.orbit_bump[0, 0, :]),
+                    time=self.orbit_bump[0, 0, :])
```

Comment:
The shaping will be handled internally by the `datatype` object if you pass `time` as an argument.
```suggestion
                self.length = ring_programs.orbit_length_program(
                    self.length_design,
                    time=self.orbit_bump[0, 0, :])
```

Reply:
I tried but unfortunately it gave me a `single` datatype.

In a test
```python
>>> ring_programs.orbit_length_program(1., time=np.array([0,1,2]))
orbit_length_program([1.])
```

Reply:
Shit.

Could you add that to the improvements issue?

Reply:
Added in the github issue